<figure>
<center>
<img src='https://www.economicas.uba.ar/wp-content/uploads/2020/08/cropped-logo_FCE.png' />
</figure>

# **Universidad de Buenos Aires**
## **Facultad de Ciencias Económicas**

### **Taller de Programación para el Análisis de Datos**

#### **Práctica Unidad 2 Primera parte**

En base al material (ejemplos y actividades) visto en clase, resuelva los siguientes ejercicios

### **Ejercicio 1**



**Cree una función denoiminada descarga que devuelva el path donde se descargó el siguiente DataFrame : https://www.kaggle.com/datasets/janiobachmann/bank-marketing-dataset Luego utilicelá para leer el dataframe con pandas (deberpa completar /NOMBREARCHIVO.csv)**

In [2]:
# Ejercicio 1.1
import kagglehub
import pandas as pd
import os

def descarga():
    """
    Descarga el dataset de Bank Marketing desde Kaggle
    y devuelve el path donde se descargó
    """
    path = kagglehub.dataset_download("janiobachmann/bank-marketing-dataset")
    print(f"Path de descarga: {path}")
    return path

# Descargar el dataset
path = descarga()

# Leer el DataFrame con pandas
# El archivo se llama bank.csv
df = pd.read_csv(os.path.join(path, "bank.csv"))

# Mostrar información básica del DataFrame
print(f"\nDataFrame cargado exitosamente")
print(f"Dimensiones: {df.shape}")
df.head()

Path de descarga: C:\Users\Alexander\.cache\kagglehub\datasets\janiobachmann\bank-marketing-dataset\versions\1

DataFrame cargado exitosamente
Dimensiones: (11162, 17)


,age,job,marital,education,default,balance,housing,loan,contact,day,month,duration,campaign,pdays,previous,poutcome,deposit
0,59,admin.,married,secondary,no,2343,yes,no,unknown,5,may,1042,1,-1,0,unknown,yes
1,56,admin.,married,secondary,no,45,no,no,unknown,5,may,1467,1,-1,0,unknown,yes
2,41,technician,married,secondary,no,1270,yes,no,unknown,5,may,1389,1,-1,0,unknown,yes
3,55,services,married,secondary,no,2476,yes,no,unknown,5,may,579,1,-1,0,unknown,yes
4,54,admin.,married,tertiary,no,184,no,no,unknown,5,may,673,2,-1,0,unknown,yes


**Describa brevemente de que se tratan los datos**

Realizar descripción

**Cree una función denominada atributos que devuelva cuantas filas y columnas tiene el dataframe, cuantas de tipo string, cuantas de tipo int, cuantas de tipo float. Aplique la función para obtener las lineas escritas**

In [3]:
def atributos(dataframe):
    """
    Devuelve información sobre la estructura del DataFrame:
    - Cantidad de filas y columnas
    - Cantidad de columnas por tipo de dato (string, int, float)
    """
    # Dimensiones del DataFrame
    num_filas, num_columnas = dataframe.shape
    
    # Contar columnas por tipo de dato
    tipos = dataframe.dtypes.value_counts()
    
    # Contar específicamente por categoría
    num_string = len(dataframe.select_dtypes(include=['object']).columns)
    num_int = len(dataframe.select_dtypes(include=['int64', 'int32']).columns)
    num_float = len(dataframe.select_dtypes(include=['float64', 'float32']).columns)
    
    # Mostrar resultados
    print("="*60)
    print("ANÁLISIS DE ATRIBUTOS DEL DATAFRAME")
    print("="*60)
    print(f"\n📊 DIMENSIONES:")
    print(f"   • Número de filas:    {num_filas:,}")
    print(f"   • Número de columnas: {num_columnas}")
    
    print(f"\n📋 COLUMNAS POR TIPO DE DATO:")
    print(f"   • Tipo String (object): {num_string}")
    print(f"   • Tipo Int:             {num_int}")
    print(f"   • Tipo Float:           {num_float}")
    
    print(f"\n🔍 DETALLE DE TIPOS:")
    for tipo, cantidad in tipos.items():
        print(f"   • {tipo}: {cantidad} columna(s)")
    
    print("="*60)
    
    # Retornar diccionario con los resultados
    return {
        'num_filas': num_filas,
        'num_columnas': num_columnas,
        'num_string': num_string,
        'num_int': num_int,
        'num_float': num_float
    }

# Aplicar la función al DataFrame
resultado = atributos(df)

ANÁLISIS DE ATRIBUTOS DEL DATAFRAME

📊 DIMENSIONES:
   • Número de filas:    11,162
   • Número de columnas: 17

📋 COLUMNAS POR TIPO DE DATO:
   • Tipo String (object): 10
   • Tipo Int:             7
   • Tipo Float:           0

🔍 DETALLE DE TIPOS:
   • object: 10 columna(s)
   • int64: 7 columna(s)


**Escriba un programa para la detección valores extremos y valores faltantes o deconocidos, con salida detallada (tener en cuenta las variables string)**



In [4]:
import numpy as np

# ============================================================================
# PROGRAMA DE DETECCIÓN DE VALORES FALTANTES Y EXTREMOS
# ============================================================================

print("="*80)
print("ANÁLISIS DE CALIDAD DE DATOS")
print("="*80)

# ----------------------------------------------------------------------------
# 1. DETECCIÓN DE VALORES FALTANTES
# ----------------------------------------------------------------------------
print("\n" + "="*80)
print("1. VALORES FALTANTES (NaN / Missing Values)")
print("="*80)

valores_faltantes = df.isna().sum()
porcentaje_faltantes = (valores_faltantes / len(df)) * 100

# Crear DataFrame con información de valores faltantes
info_faltantes = pd.DataFrame({
    'Columna': valores_faltantes.index,
    'Cantidad_Faltantes': valores_faltantes.values,
    'Porcentaje_%': porcentaje_faltantes.values
})

# Filtrar solo las que tienen valores faltantes
info_faltantes = info_faltantes[info_faltantes['Cantidad_Faltantes'] > 0]

if len(info_faltantes) > 0:
    print(f"\n⚠️  Se encontraron valores faltantes en {len(info_faltantes)} columna(s):\n")
    print(info_faltantes.to_string(index=False))
else:
    print("\n✅ No se encontraron valores faltantes en ninguna columna")

print(f"\nTotal de valores faltantes en el dataset: {valores_faltantes.sum():,}")

# ----------------------------------------------------------------------------
# 2. DETECCIÓN DE VALORES EXTREMOS (OUTLIERS) - VARIABLES NUMÉRICAS
# ----------------------------------------------------------------------------
print("\n" + "="*80)
print("2. VALORES EXTREMOS (OUTLIERS) - Método de Tukey (RIQ)")
print("="*80)

# Seleccionar solo columnas numéricas
columnas_numericas = df.select_dtypes(include=[np.number]).columns

print(f"\nAnalizando {len(columnas_numericas)} variable(s) numérica(s)...\n")

outliers_info = []

for col in columnas_numericas:
    # Calcular estadísticos
    Q1 = df[col].quantile(0.25)
    Q3 = df[col].quantile(0.75)
    RIQ = Q3 - Q1
    
    # Límites de Tukey
    limite_inferior = Q1 - 1.5 * RIQ
    limite_superior = Q3 + 1.5 * RIQ
    
    # Detectar outliers
    outliers = df[(df[col] < limite_inferior) | (df[col] > limite_superior)][col]
    num_outliers = len(outliers)
    porcentaje_outliers = (num_outliers / len(df)) * 100
    
    if num_outliers > 0:
        outliers_info.append({
            'Columna': col,
            'Num_Outliers': num_outliers,
            'Porcentaje_%': f"{porcentaje_outliers:.2f}",
            'Límite_Inferior': f"{limite_inferior:.2f}",
            'Límite_Superior': f"{limite_superior:.2f}",
            'Min_Outlier': f"{outliers.min():.2f}",
            'Max_Outlier': f"{outliers.max():.2f}"
        })
        
        print(f"📊 {col}:")
        print(f"   • Outliers detectados: {num_outliers} ({porcentaje_outliers:.2f}%)")
        print(f"   • Rango normal: [{limite_inferior:.2f}, {limite_superior:.2f}]")
        print(f"   • Rango de outliers: [{outliers.min():.2f}, {outliers.max():.2f}]")
        print()

if len(outliers_info) == 0:
    print("✅ No se detectaron outliers en las variables numéricas")
else:
    print(f"\n⚠️  RESUMEN: Se detectaron outliers en {len(outliers_info)} variable(s)")

# ----------------------------------------------------------------------------
# 3. ANÁLISIS DE VARIABLES STRING (CATEGÓRICAS)
# ----------------------------------------------------------------------------
print("\n" + "="*80)
print("3. ANÁLISIS DE VARIABLES CATEGÓRICAS (STRING)")
print("="*80)

# Seleccionar columnas de tipo object (string)
columnas_string = df.select_dtypes(include=['object']).columns

print(f"\nAnalizando {len(columnas_string)} variable(s) categórica(s)...\n")

valores_inusuales = []

for col in columnas_string:
    # Contar valores únicos
    conteo_valores = df[col].value_counts()
    total_valores = len(df[col])
    
    print(f"📝 {col}:")
    print(f"   • Valores únicos: {len(conteo_valores)}")
    print(f"   • Valor más frecuente: '{conteo_valores.index[0]}' ({conteo_valores.iloc[0]} casos)")
    
    # Detectar valores poco frecuentes (< 1% del total)
    valores_raros = conteo_valores[conteo_valores < (total_valores * 0.01)]
    
    if len(valores_raros) > 0:
        print(f"   ⚠️  Valores poco frecuentes (<1%): {len(valores_raros)}")
        # Mostrar algunos ejemplos
        ejemplos = list(valores_raros.head(3).index)
        print(f"      Ejemplos: {ejemplos}")
        valores_inusuales.append(col)
    
    # Detectar valores "unknown" o similares
    valores_desconocidos = df[col].isin(['unknown', 'Unknown', 'UNKNOWN', 'N/A', 'NA', '', ' '])
    num_desconocidos = valores_desconocidos.sum()
    
    if num_desconocidos > 0:
        print(f"   ⚠️  Valores desconocidos/vacíos: {num_desconocidos} ({(num_desconocidos/len(df)*100):.2f}%)")
    
    print()

if len(valores_inusuales) > 0:
    print(f"\n⚠️  Variables con valores poco frecuentes: {valores_inusuales}")
else:
    print("\n✅ No se detectaron valores inusuales en las variables categóricas")

# ----------------------------------------------------------------------------
# RESUMEN FINAL
# ----------------------------------------------------------------------------
print("\n" + "="*80)
print("RESUMEN DEL ANÁLISIS DE CALIDAD")
print("="*80)
print(f"✓ Valores faltantes: {valores_faltantes.sum():,} en total")
print(f"✓ Variables numéricas con outliers: {len(outliers_info)}")
print(f"✓ Variables categóricas con valores inusuales: {len(valores_inusuales)}")
print("="*80)

ANÁLISIS DE CALIDAD DE DATOS

1. VALORES FALTANTES (NaN / Missing Values)

✅ No se encontraron valores faltantes en ninguna columna

Total de valores faltantes en el dataset: 0

2. VALORES EXTREMOS (OUTLIERS) - Método de Tukey (RIQ)

Analizando 7 variable(s) numérica(s)...

📊 age:
   • Outliers detectados: 171 (1.53%)
   • Rango normal: [6.50, 74.50]
   • Rango de outliers: [75.00, 95.00]

📊 balance:
   • Outliers detectados: 1055 (9.45%)
   • Rango normal: [-2257.00, 4087.00]
   • Rango de outliers: [-6847.00, 81204.00]

📊 duration:
   • Outliers detectados: 636 (5.70%)
   • Rango normal: [-399.00, 1033.00]
   • Rango de outliers: [1034.00, 3881.00]

📊 campaign:
   • Outliers detectados: 601 (5.38%)
   • Rango normal: [-2.00, 6.00]
   • Rango de outliers: [7.00, 63.00]

📊 pdays:
   • Outliers detectados: 2750 (24.64%)
   • Rango normal: [-33.62, 53.38]
   • Rango de outliers: [55.00, 854.00]

📊 previous:
   • Outliers detectados: 1258 (11.27%)
   • Rango normal: [-1.50, 2.50]
   • Ran

**Escriba un programa que devuelva la cantidad de casos que hay por categoria en cada variable string**

In [ ]:
# ============================================================================
# ANÁLISIS DE FRECUENCIAS POR CATEGORÍA EN VARIABLES STRING
# ============================================================================

print("="*80)
print("CANTIDAD DE CASOS POR CATEGORÍA - VARIABLES CATEGÓRICAS")
print("="*80)

# Seleccionar columnas de tipo string (object)
columnas_categoricas = df.select_dtypes(include=['object']).columns

print(f"\nSe encontraron {len(columnas_categoricas)} variable(s) categórica(s)\n")

# Analizar cada variable categórica
for col in columnas_categoricas:
    print("="*80)
    print(f"VARIABLE: {col}")
    print("="*80)
    
    # Contar frecuencias
    frecuencias = df[col].value_counts()
    porcentajes = (df[col].value_counts(normalize=True) * 100).round(2)
    
    # Crear DataFrame con información detallada
    info_categorias = pd.DataFrame({
        'Categoría': frecuencias.index,
        'Cantidad': frecuencias.values,
        'Porcentaje_%': porcentajes.values
    })
    
    # Mostrar información
    print(f"\nNúmero de categorías únicas: {len(frecuencias)}")
    print(f"Total de casos: {frecuencias.sum():,}\n")
    print(info_categorias.to_string(index=False))
    print()

# Resumen general
print("="*80)
print("RESUMEN GENERAL")
print("="*80)
for col in columnas_categoricas:
    num_categorias = df[col].nunique()
    print(f"• {col}: {num_categorias} categoría(s) única(s)")
print("="*80)

**Escriba un programa que cree una nueva variable denominada TIPO_SALDO, según las siguientes reglas:**

- Si el saldo es negativo -> 'NEGATIVO'
- Si el saldo es nulo -> 'NULO'
- Si el saldo es positivo -> 'POSITIVO'

**y que al final devuelva la cantidad de casos por categoría**

In [ ]:
# ============================================================================
# CREACIÓN DE VARIABLE TIPO_SALDO
# ============================================================================

print("="*80)
print("CREACIÓN DE VARIABLE: TIPO_SALDO")
print("="*80)

# Nota: Asumiendo que la columna de saldo se llama 'balance' 
# (ajustar según el nombre real en el dataset)

# Verificar si existe la columna 'balance'
if 'balance' in df.columns:
    columna_saldo = 'balance'
else:
    # Buscar columna que contenga 'saldo' o 'balance' en el nombre
    posibles_columnas = [col for col in df.columns if 'saldo' in col.lower() or 'balance' in col.lower()]
    if posibles_columnas:
        columna_saldo = posibles_columnas[0]
        print(f"ℹ️  Usando columna: {columna_saldo}\n")
    else:
        print("⚠️  No se encontró columna de saldo. Ajustar el nombre de la columna.")
        columna_saldo = None

if columna_saldo:
    # Crear la nueva variable TIPO_SALDO aplicando las reglas
    def clasificar_saldo(valor):
        if pd.isna(valor) or valor == 0:
            return 'NULO'
        elif valor < 0:
            return 'NEGATIVO'
        else:  # valor > 0
            return 'POSITIVO'
    
    # Aplicar la función a cada valor
    df['TIPO_SALDO'] = df[columna_saldo].apply(clasificar_saldo)
    
    print("✅ Variable TIPO_SALDO creada exitosamente\n")
    
    # Contar casos por categoría
    print("="*80)
    print("CANTIDAD DE CASOS POR CATEGORÍA")
    print("="*80)
    
    frecuencias = df['TIPO_SALDO'].value_counts()
    porcentajes = (df['TIPO_SALDO'].value_counts(normalize=True) * 100).round(2)
    
    resultado = pd.DataFrame({
        'Tipo_Saldo': frecuencias.index,
        'Cantidad': frecuencias.values,
        'Porcentaje_%': porcentajes.values
    })
    
    print(f"\nTotal de casos: {len(df):,}\n")
    print(resultado.to_string(index=False))
    print("\n" + "="*80)
    
    # Mostrar algunos ejemplos
    print("\nEjemplos de registros por categoría:")
    print("="*80)
    for tipo in df['TIPO_SALDO'].unique():
        print(f"\n{tipo}:")
        ejemplos = df[df['TIPO_SALDO'] == tipo][[columna_saldo, 'TIPO_SALDO']].head(3)
        print(ejemplos.to_string(index=False))

**Escriba un programa que cree una nueva variable denominada RIESGO, según las siguientes reglas:**

- Si el tipo de saldo es nulo o negativo y alguna vez entró en default -> 'ALTO'
- Si el tipo de saldo es positivo y alguna vez entró en default -> 'MEDIO'
- Si nunca entró en Default -> 'BAJO'
- En cualquier otro caso-> 'INDEFINIDO'

**y que al final devuelva la cantidad de casos por categoría**

In [ ]:
# ============================================================================
# CREACIÓN DE VARIABLE RIESGO
# ============================================================================

print("="*80)
print("CREACIÓN DE VARIABLE: RIESGO")
print("="*80)

# Verificar que existe TIPO_SALDO (creada en ejercicio anterior)
if 'TIPO_SALDO' not in df.columns:
    print("⚠️  Error: Primero debe ejecutar el ejercicio anterior para crear TIPO_SALDO")
else:
    # Verificar si existe la columna 'default'
    if 'default' in df.columns:
        columna_default = 'default'
    else:
        # Buscar columna que contenga 'default' en el nombre
        posibles_columnas = [col for col in df.columns if 'default' in col.lower()]
        if posibles_columnas:
            columna_default = posibles_columnas[0]
            print(f"ℹ️  Usando columna: {columna_default}\n")
        else:
            print("⚠️  No se encontró columna de default. Ajustar el nombre de la columna.")
            columna_default = None
    
    if columna_default:
        # Crear la variable RIESGO aplicando las reglas
        def clasificar_riesgo(row):
            tipo_saldo = row['TIPO_SALDO']
            default = row[columna_default]
            
            # Normalizar valores de default (puede venir como 'yes'/'no' o 1/0)
            if isinstance(default, str):
                entro_default = default.lower() in ['yes', 'si', 'sí', 'y', '1', 'true']
            else:
                entro_default = bool(default)
            
            # Aplicar reglas
            if (tipo_saldo in ['NULO', 'NEGATIVO']) and entro_default:
                return 'ALTO'
            elif tipo_saldo == 'POSITIVO' and entro_default:
                return 'MEDIO'
            elif not entro_default:
                return 'BAJO'
            else:
                return 'INDEFINIDO'
        
        # Aplicar la función a cada fila
        df['RIESGO'] = df.apply(clasificar_riesgo, axis=1)
        
        print("✅ Variable RIESGO creada exitosamente\n")
        
        # Contar casos por categoría
        print("="*80)
        print("CANTIDAD DE CASOS POR CATEGORÍA - RIESGO")
        print("="*80)
        
        frecuencias = df['RIESGO'].value_counts()
        porcentajes = (df['RIESGO'].value_counts(normalize=True) * 100).round(2)
        
        resultado = pd.DataFrame({
            'Nivel_Riesgo': frecuencias.index,
            'Cantidad': frecuencias.values,
            'Porcentaje_%': porcentajes.values
        })
        
        print(f"\nTotal de casos: {len(df):,}\n")
        print(resultado.to_string(index=False))
        print("\n" + "="*80)
        
        # Tabla cruzada para entender la relación
        print("\nTABLA CRUZADA: TIPO_SALDO vs RIESGO")
        print("="*80)
        tabla_cruzada = pd.crosstab(df['TIPO_SALDO'], df['RIESGO'], margins=True)
        print(tabla_cruzada)
        print()
        
        # Mostrar algunos ejemplos
        print("\nEjemplos de registros por nivel de riesgo:")
        print("="*80)
        for nivel in ['ALTO', 'MEDIO', 'BAJO', 'INDEFINIDO']:
            if nivel in df['RIESGO'].values:
                print(f"\n{nivel}:")
                ejemplos = df[df['RIESGO'] == nivel][['TIPO_SALDO', columna_default, 'RIESGO']].head(3)
                print(ejemplos.to_string(index=False))

**Seleccione dos muestras aletorias: una con reemplazo y otra sin reemplazo de tamaño 1000 cada una.**

**Al final, concatene ambas muestras en un nuevo DataFrame y verifique si tiene casos duplicados. Si hay casos duplicados, eliminelos y describa las dimensiones finales**

In [ ]:
# ============================================================================
# MUESTREO ALEATORIO Y GESTIÓN DE DUPLICADOS
# ============================================================================

print("="*80)
print("MUESTREO ALEATORIO Y ANÁLISIS DE DUPLICADOS")
print("="*80)

# Tamaño de cada muestra
tamanio_muestra = 1000
seed = 42  # Semilla para reproducibilidad

print(f"\nDataFrame original: {len(df):,} filas\n")

# ----------------------------------------------------------------------------
# 1. MUESTRA SIN REEMPLAZO (replace=False)
# ----------------------------------------------------------------------------
print("1. MUESTRA SIN REEMPLAZO")
print("-" * 80)

# Verificar si hay suficientes datos para muestra sin reemplazo
if len(df) >= tamanio_muestra:
    muestra_sin_reemplazo = df.sample(n=tamanio_muestra, replace=False, random_state=seed)
    print(f"✅ Muestra sin reemplazo creada: {len(muestra_sin_reemplazo):,} filas")
    print(f"   • Cada registro aparece máximo 1 vez")
else:
    print(f"⚠️  No hay suficientes datos. Se requieren al menos {tamanio_muestra} filas")
    muestra_sin_reemplazo = df.sample(n=len(df), replace=False, random_state=seed)
    print(f"   Muestra creada con {len(muestra_sin_reemplazo):,} filas (todas disponibles)")

# ----------------------------------------------------------------------------
# 2. MUESTRA CON REEMPLAZO (replace=True)
# ----------------------------------------------------------------------------
print("\n2. MUESTRA CON REEMPLAZO")
print("-" * 80)

muestra_con_reemplazo = df.sample(n=tamanio_muestra, replace=True, random_state=seed)
print(f"✅ Muestra con reemplazo creada: {len(muestra_con_reemplazo):,} filas")
print(f"   • Los registros pueden repetirse")

# Verificar duplicados dentro de la muestra con reemplazo
duplicados_interna = muestra_con_reemplazo.duplicated().sum()
print(f"   • Duplicados dentro de esta muestra: {duplicados_interna}")

# ----------------------------------------------------------------------------
# 3. CONCATENAR AMBAS MUESTRAS
# ----------------------------------------------------------------------------
print("\n3. CONCATENACIÓN DE MUESTRAS")
print("-" * 80)

# Concatenar las dos muestras
df_concatenado = pd.concat([muestra_sin_reemplazo, muestra_con_reemplazo], ignore_index=True)

print(f"✅ Muestras concatenadas")
print(f"   • Dimensiones esperadas: ({len(muestra_sin_reemplazo) + len(muestra_con_reemplazo)}, {df_concatenado.shape[1]})")
print(f"   • Dimensiones obtenidas: {df_concatenado.shape}")

# ----------------------------------------------------------------------------
# 4. VERIFICAR DUPLICADOS
# ----------------------------------------------------------------------------
print("\n4. ANÁLISIS DE DUPLICADOS")
print("-" * 80)

# Contar duplicados
num_duplicados = df_concatenado.duplicated().sum()
print(f"\n📊 Duplicados encontrados: {num_duplicados:,} fila(s)")

if num_duplicados > 0:
    print(f"   • Porcentaje de duplicados: {(num_duplicados/len(df_concatenado)*100):.2f}%")
    
    # Mostrar algunos ejemplos de duplicados
    print("\n   Ejemplos de registros duplicados:")
    duplicados_ejemplos = df_concatenado[df_concatenado.duplicated(keep=False)].head(6)
    print(f"   (Mostrando primeras 6 filas de duplicados)\n")
    print(duplicados_ejemplos)

# ----------------------------------------------------------------------------
# 5. ELIMINAR DUPLICADOS
# ----------------------------------------------------------------------------
print("\n5. ELIMINACIÓN DE DUPLICADOS")
print("-" * 80)

dimensiones_antes = df_concatenado.shape
df_final = df_concatenado.drop_duplicates()
dimensiones_despues = df_final.shape

print(f"\n✅ Duplicados eliminados")
print(f"\n📊 COMPARACIÓN DE DIMENSIONES:")
print(f"   • Antes de eliminar:  {dimensiones_antes[0]:,} filas × {dimensiones_antes[1]} columnas")
print(f"   • Después de eliminar: {dimensiones_despues[0]:,} filas × {dimensiones_despues[1]} columnas")
print(f"   • Filas eliminadas:    {dimensiones_antes[0] - dimensiones_despues[0]:,}")

# Verificación final
print(f"\n🔍 VERIFICACIÓN FINAL:")
print(f"   • Duplicados restantes: {df_final.duplicated().sum()}")
print(f"   • Dataset final limpio: {len(df_final):,} registros únicos")

print("\n" + "="*80)
print("PROCESO COMPLETADO")
print("="*80)

# Guardar información del DataFrame final
print(f"\nDataFrame final 'df_final' disponible con {len(df_final):,} registros")

### **Ejercicio 2**

2.1 Escribir un programa con los siguientes requisitos:

Descargue el DataFrame titanic.csv que contiene información sobre los pasajeros del Titanic: https://aprendeconalf.es/docencia/python/ejercicios/soluciones/pandas/titanic.csv y generar un DataFrame con el nombre DATA.


2.2 Escribir un programa con los siguientes requisitos:

Mostrar por pantalla las dimensiones del DataFrame, el número de casos que contiene y columnas, los nombres de sus columnas, la cantidad de columnas por tipo de dato.


2.3 Escribir un programa con los siguientes requisitos:

Mostrar por pantalla los datos del pasajero con identificador 148.


2.4 Escribir un programa con los siguientes requisitos:

Mostrar por pantalla los nombres de las personas que iban en primera clase ordenadas alfabéticamente.

2.5 Escribir un programa con los siguientes requisitos:

Mostrar por pantalla el porcentaje de personas que sobrevivieron y murieron.


2.6 Escribir un programa con los siguientes requisitos:

Mostrar por pantalla el porcentaje de personas que sobrevivieron en cada clase.

2.7 Escribir un programa con los siguientes requisitos:

Eliminar del DataFrame los pasajeros con edad desconocida, mostrando cuantos casos se eliminaron

2.8 Escribir un programa con los siguientes requisitos:

Crear una nueva columna booleana para ver si el pasajero era menor de edad o no. Mostrar por pantalla el porcentaje de menores y mayores de edad que sobrevivieron en cada clase.